### ISSUE TITLE
Python sample speech_recognition_with_push_stream blocks long time

### ISSUE THREAD

**Issue by PatrikValo** [Azure-Samples/cognitive-services-speech-sdk#1199]
**Describe the bug**
Python speech sample function `speech_recognition_with_push_stream` contains line `speech_recognizer.stop_continuous_recognition()`, which blocks for 10s. I guess that 10s (10 000ms) period could be a timeout, which probably expires because speech recognizer doesn't detect, that stream has been already closed in previous step and method is blocking and waiting until stream is closed.
https://github.com/Azure-Samples/cognitive-services-speech-sdk/blob/ac8c9899c4c741a836fbcd01d66b9f4fd1d06cec/samples/python/console/speech_sample.py#L453-L457
I found out and tried these scenarios:
- when I add some debug prints
```python
stream.close()
print("Before...")
start = time.time()
speech_recognizer.stop_continuous_recognition()
print(f"After {time.time() - start}...")
```
output is
```shell
Before...
RECOGNIZED: SpeechRecognitionEventArgs(session_id=<session_id>, result=SpeechRecognitionResult(result_id=<result_id>, text="", reason=ResultReason.RecognizedSpeech))
After 10.02142333984375...
```

- when I add `time.sleep(1)` after closing the stream, it is without 10s delay
```python
stream.close()
time.sleep(1)
print("Before...")
start = time.time()
speech_recognizer.stop_continuous_recognition()
print(f"After {time.time() - start}...")
```
```shell
RECOGNIZED: SpeechRecognitionEventArgs(session_id=<session_id>, result=SpeechRecognitionResult(result_id=<result_id>, text="", reason=ResultReason.RecognizedSpeech))
CANCELED SpeechRecognitionCanceledEventArgs(session_id=<session_id>, result=SpeechRecognitionResult(result_id=<result_id>, text="", reason=ResultReason.Canceled))
SESSION STOPPED SessionEventArgs(session_id=<session_id>)
Before...
After 0.0006592273712158203...
```
- when I replace `speech_recognizer.stop_continuous_recognition()` by async version, everything works fine without 10s delay
```python
stream.close()
print("Before...")
start = time.time()
speech_recognizer.stop_continuous_recognition_async().get()
print(f"After {time.time() - start}...")
```
```shell
Before...
RECOGNIZED: SpeechRecognitionEventArgs(session_id=<session_id>, result=SpeechRecognitionResult(result_id=<result_id>, text="", reason=ResultReason.RecognizedSpeech))
SESSION STOPPED SessionEventArgs(session_id=<session_id>)
After 0.598764181137085...
```

**To Reproduce**
1. Run function `speech_recognition_with_push_stream`
https://github.com/Azure-Samples/cognitive-services-speech-sdk/blob/ac8c9899c4c741a836fbcd01d66b9f4fd1d06cec/samples/python/console/speech_sample.py#L417
2. It takes 10 000 ms to stop execution of the `stop_continuous_recognition` method

**Expected behavior**
 `stop_continuous_recognition` should not block so long.

**Version of the Cognitive Services Speech SDK**
azure-cognitiveservices-speech version 1.18.0

**Platform, Operating System, and Programming Language**
 - OS: Ubuntu 18.04
 - Hardware - x64
 - Programming language: Python

**Comment by pankopon** [comment]
Hi, `stop_continuous_recognition` synchronously terminates recognition, which may block depending on when the call is made and the state of operations at the moment, while `stop_continuous_recognition_async` does it asynchronously. We can improve the sample to demonstrate the use of `stop_continuous_recognition_async` as well.

**Comment by PatrikValo** [comment]
Hi, thank you very much for fast answer. 
Yes, I understand but what I would like to say is, that blocking time in your example (10 000 ms) is too big and probably it can indicates some problem. 

I enabled the logging in your sample `speech_recognition_with_push_stream` (without any additional modification of your code). In the log I found this lines:
```
[741860]: 1825ms SPX_DBG_TRACE_SCOPE_EXIT:  usp_reco_engine_adapter.cpp:1920 CSpxUspRecoEngineAdapter::OnTurnEnd ... DONE!
[788145]: 11521ms SPX_DBG_TRACE_VERBOSE:  audio_stream_session.cpp:771 CSpxAudioStreamSession::WaitForIdle m_cv.wait_for, m_sessionState: 0, m_recoKind: 0, m_sessionStarted: 1, m_sessionStopped: 0
[788145]: 11522ms SPX_TRACE_WARNING: audio_stream_session.cpp:780 [0x21bde00]CSpxAudioStreamSession::WaitForIdle: Timeout happened during waiting for Idle
```
I think that the most interesting part is time between two consecutive logs line ~10 000ms and also message **Timeout happened during waiting for Idle**.

I try to rewrite your sample with taking inspiration from another samples. I add `stop_cb` function and `done` flag. And `stop_continuous_recognition` method is called if `done` flag is equal to value `True`.
```python
def speech_recognition_with_push_stream():
    """gives an example how to use a push audio stream to recognize speech from a custom audio
    source"""
    speech_config = speechsdk.SpeechConfig(subscription=speech_key, region=service_region)
    speech_config.set_property(speechsdk.PropertyId.Speech_LogFilename, "log.txt")

    # setup the audio stream
    stream = speechsdk.audio.PushAudioInputStream()
    audio_config = speechsdk.audio.AudioConfig(stream=stream)

    done = False

    def stop_cb(evt):
        """callback that signals to stop continuous recognition upon receiving an event `evt`"""
        print('CLOSING on {}'.format(evt))
        nonlocal done
        done = True

    # instantiate the speech recognizer with push stream input
    speech_recognizer = speechsdk.SpeechRecognizer(speech_config=speech_config, audio_config=audio_config)

    # Connect callbacks to the events fired by the speech recognizer
    speech_recognizer.recognizing.connect(lambda evt: print('RECOGNIZING: {}'.format(evt)))
    speech_recognizer.recognized.connect(lambda evt: print('RECOGNIZED: {}'.format(evt)))
    speech_recognizer.session_started.connect(lambda evt: print('SESSION STARTED: {}'.format(evt)))
    speech_recognizer.session_stopped.connect(lambda evt: print('SESSION STOPPED {}'.format(evt)))
    speech_recognizer.canceled.connect(lambda evt: print('CANCELED {}'.format(evt)))
    speech_recognizer.session_stopped.connect(stop_cb)
    speech_recognizer.canceled.connect(stop_cb)

    # The number of bytes to push per buffer
    n_bytes = 3200
    wav_fh = wave.open(weatherfilename)

    # start continuous speech recognition
    speech_recognizer.start_continuous_recognition()

    # start pushing data until all data has been read from the file
    try:
        while(True):
            frames = wav_fh.readframes(n_bytes // 2)
            print('read {} bytes'.format(len(frames)))
            if not frames:
                break

            stream.write(frames)
            time.sleep(.1)
    finally:
        # stop recognition and clean up
        wav_fh.close()
        stream.close()
        print("Before...")
        start = time.time()
        while not done:
            time.sleep(0.5)
        speech_recognizer.stop_continuous_recognition()
        print(f"After {time.time() - start}...")
```
and in this case, blocking time (+ waiting time until done flag is True) was only 501ms.
```shell
After 0.5018365383148193...
```
When I enable logging, timeout message disappears.

**So my final question is: Is your sample with synchronous `stop_continuous_recognition` method implemented correctly and blocking time 10 000ms is correct behaviour? Or method `stop_continuous_recognition` should be called after `cancel` or `stop` callback execution?**

Thank you very much for your time and patience with me.

**Additional note**
For example:
```python
speech_recognizer.stop_continuous_recognition_async().get()
```
async method return future object, which has method `get`. If I understand it correctly, `get` method blocks in same manner like `stop_continuous_recognition`. It blocks depending on when the call is made and the state of operations at the moment. It is blocking until the  termination of the recognition is done. And it blocks in your example only 598ms.

**Comment by pankopon** [comment]
Thanks for the feedback and improvement suggestions. The 10 sec wait is indeed due the recognizer/service not returning to an idle state before an SDK internal timeout, this is controlled in native C++ implementation. Sometimes it may or may not trigger in the same use case depending on just small timing differences in the order of internal operations, though in this case it is admittedly consistent. I'll take a look at the root cause, but at the very least we can improve the samples in this regard.

**Comment by pankopon** [comment]
This will be fixed in the Speech SDK 1.20 release, expected late Jan / early Feb. Internal work item ref. 3712714.

**Comment by pankopon** [comment]
Fixed in the Speech SDK release 1.21.0 which is available now.
